### PR TITLE
Custom remote keys

### PR DESCRIPTION
**Pull request by RamonGilabert**
I've added the code to check if the user info of any key it's full with a mapper.remote.key value, if it's not, or if it has a nil value, named "value", then it puts the default key, in the tests you'll se an error I need to fix that is, check if the user has a key with user info, if it has a user info key, then you don't need to check if it's nil the default key but the userInfo one.

### CODE REVIEW

**Comment by NSElvis** [comment]
:scream:

**Comment by NSElvis** [comment]
:bow:

File: Source/NSManagedObject+HYPPropertyMapper.m [hunk 1]
```diff
@@ -129,28 +129,51 @@ - (NSDictionary *)hyp_dictionary
...
     for (id propertyDescription in [self.entity properties]) {
         if ([propertyDescription isKindOfClass:[NSAttributeDescription class]]) {
-            NSAttributeDescription *attributeDescription = (NSAttributeDescription *)propertyDescription;
-            id value = [self valueForKey:[attributeDescription name]];
-            NSMutableString *key = [[[propertyDescription name] hyp_remoteString] mutableCopy];
-
-            BOOL nilOrNullValue = (!value || [value isKindOfClass:[NSNull class]]);
-            if (nilOrNullValue) {
-                mutableDictionary[key] = [NSNull null];
+            if ([[propertyDescription userInfo] objectForKey:@"mapper.remote.key"] && ![[[propertyDescription userInfo] objectForKey:@"mapper.remote.key"] isEqualToString:@"value"]) {
```
Comment (zenangst) [inline Source/NSManagedObject+HYPPropertyMapper.m]: `mapper.remote.key` could probably be made into a constant

**Comment by zenangst** [comment]
Would you mind adding some tests? :)

File: NSManagedObject-HYPPropertyMapperTests/NSManagedObject_HYPPropertyMapperTests.m [hunk 2]
```diff
@@ -112,68 +112,119 @@ - (void)tearDown
...
 {
     NSDictionary *dictionary = [self.testUser hyp_dictionary];
 
-    XCTAssertTrue([dictionary[@"age"] isKindOfClass:[NSNumber class]]);
+    NSArray *arraySortedKeys = [self sortArrayOfKeysFromDictionary:dictionary];
+
+    XCTAssertTrue([dictionary[arraySortedKeys[0]] isKindOfClass:[NSNumber class]]);
 
-    XCTAssertTrue([dictionary[@"birth_date"] isKindOfClass:[NSDate class]]);
+    XCTAssertTrue([dictionary[arraySortedKeys[1]] isKindOfClass:[NSDate class]]);
```
Comment (NSElvis) [inline NSManagedObject-HYPPropertyMapperTests/NSManagedObject_HYPPropertyMapperTests.m]: Hey @RamonGilabert! I'm afraid I don't share the your same opinion here, I can't see how

```objc
XCTAssertTrue([dictionary[arraySortedKeys[1]] isKindOfClass:[NSDate class]]);
```

Is easier to understand than

```objc
XCTAssertTrue([dictionary[@"birth_date"] isKindOfClass:[NSDate class]]);
```

I don't think tests needed to be refactored 😕

File: Source/NSManagedObject+HYPPropertyMapper.m [hunk 3]
```diff
@@ -2,6 +2,8 @@
 
 #import "NSString+HYPNetworking.h"
 
+static NSString * const HYPPropertyMapperRemoteKey = @"mapper.remote.key";
```
Comment (NSElvis) [inline Source/NSManagedObject+HYPPropertyMapper.m]: Nice!

File: Source/NSManagedObject+HYPPropertyMapper.m [hunk 4]
```diff
@@ -129,28 +131,51 @@ - (NSDictionary *)hyp_dictionary
...
-            id value = [self valueForKey:[attributeDescription name]];
-            NSMutableString *key = [[[propertyDescription name] hyp_remoteString] mutableCopy];
-
-            BOOL nilOrNullValue = (!value || [value isKindOfClass:[NSNull class]]);
-            if (nilOrNullValue) {
-                mutableDictionary[key] = [NSNull null];
+            if ([[propertyDescription userInfo] objectForKey:HYPPropertyMapperRemoteKey] && ![[[propertyDescription userInfo] objectForKey:HYPPropertyMapperRemoteKey] isEqualToString:@"value"]) {
+                NSAttributeDescription *attributeDescription = (NSAttributeDescription *)propertyDescription;
+                id value = [self valueForKey:[attributeDescription name]];
+                NSMutableString *key = [[[[propertyDescription userInfo] objectForKey:@"mapper.remote.key"] hyp_remoteString] mutableCopy];
```
Comment (NSElvis) [inline Source/NSManagedObject+HYPPropertyMapper.m]: You might one to use your constant here `HYPPropertyMapperRemoteKey`

File: NSManagedObject-HYPPropertyMapperTests/NSManagedObject_HYPPropertyMapperTests.m (hunk 2 above)
Comment (RamonGilabert) [inline NSManagedObject-HYPPropertyMapperTests/NSManagedObject_HYPPropertyMapperTests.m]: So the point to do this is that, if you add a key in the user info, the array with key "age" won't exist anymore, it's going to be age_of_person for instance, then I had to do all this code, I guess there's a better way. Also, I want to tell that, for some reason, when I call hyp-remoteString, it doesn't add _ when it found a space for some reason.

Comment (NSElvis) [inline NSManagedObject-HYPPropertyMapperTests/NSManagedObject_HYPPropertyMapperTests.m]: @RamonGilabert

I understand, updating the tests with the new added values hasn't been a problem so far for me, but maybe it makes it harder for collaborators.

>Also, I want to tell that, for some reason, when I call hyp-remoteString, it doesn't add _ when it found a space for some reason.

That looks like a bug in [NSString-HYPNetworking](https://github.com/hyperoslo/NSString-HYPNetworking), would you mind opening an issue? :smile: :heart:

**Comment by NSElvis** [comment]
I guess we will continue this here

https://github.com/hyperoslo/NSManagedObject-HYPPropertyMapper/pull/42

**Pull request closed without merging**